services/httpd/handler.go

```diff
@@ -993,8 +1017,437 @@ func fixLiterals(node influxql.Node) {
+	if len(brd.RetentionRules) > 0 {
+		dur = time.Second * time.Duration(brd.RetentionRules[0].EverySeconds)
+		sgDur = time.Duration(brd.RetentionRules[0].ShardGroupDurationSeconds) * time.Second
+	} else {
+		dur = meta.DefaultRetentionPolicyDuration
+		// This will get set to default in normalisedShardDuration()
+		// called by CreateRetentionPolicy
+		sgDur = 0
+	}
+	rf := meta.DefaultRetentionPolicyReplicaN
+	spec := meta.RetentionPolicySpec{
+		Name:               rp,
+		Duration:           &dur,
+		ReplicaN:           &rf,
+		ShardGroupDuration: sgDur,
+	}
+	dbi := h.MetaClient.Database(db)
+
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, fmt.Sprintf("buckets - user is required to to create bucket %q", brd.Name), http.StatusForbidden)
+			return
+		}
+
+		if dbi == nil {
+			if err := h.QueryAuthorizer.AuthorizeCreateDatabase(user); err != nil {
+				h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to create %q: %s", user.ID(), brd.Name, err.Error()), http.StatusForbidden)
+				return
+			}
+		} else {
+			if err := h.QueryAuthorizer.AuthorizeCreateRetentionPolicy(user, db); err != nil {
+				h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to create %q: %s", user.ID(), brd.Name, err.Error()), http.StatusForbidden)
+				return
+			}
+		}
+	}
+
+	if dbi == nil {
+		if _, err = h.MetaClient.CreateDatabaseWithRetentionPolicy(db, &spec); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - cannot create bucket %q: %s", brd.Name, err.Error()), http.StatusBadRequest)
+			return
+		}
+	} else {
+		if _, err = h.MetaClient.CreateRetentionPolicy(db, &spec, false); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - cannot create bucket %q: %s", brd.Name, err.Error()), http.StatusBadRequest)
+			return
+		}
+	}
+}
+
+func (h *Handler) serveBucketDeleteV2(w http.ResponseWriter, r *http.Request, user meta.User) {
+	db := r.URL.Query().Get(":db")
+	rp := r.URL.Query().Get(":rp")
+	id := fmt.Sprintf("%s/%s", db, rp)
+
+	if h.Config.AuthEnabled {
+		if user == nil {
+			h.httpError(w, fmt.Sprintf("delete bucket - user is required to delete from database %q", db), http.StatusForbidden)
+			return
+		}
+		if err := h.QueryAuthorizer.AuthorizeCreateRetentionPolicy(user, db); err != nil {
+			h.httpError(w, fmt.Sprintf("buckets - %q is not authorized to modify %q: %s", user.ID(), id, err.Error()), http.StatusForbidden)
+			return
+		}
+	}
+
+	if dbi := h.MetaClient.Database(db); dbi == nil {
+		h.httpError(w, fmt.Sprintf("delete bucket - not found: %q", id), http.StatusNotFound)
+		return
+	} else if rpi := dbi.RetentionPolicy(rp); rpi == nil {
+		h.httpError(w, fmt.Sprintf("delete bucket - not found: %q", id), http.StatusNotFound)
+		return
+	} else if err := h.Store.DeleteRetentionPolicy(db, rp); err != nil {
+		h.httpError(w, fmt.Sprintf("delete bucket %q: %s", id, err.Error()), http.StatusBadRequest)
+		return
+	} else if err := h.MetaClient.DropRetentionPolicy(db, rp); err != nil {
```

Comment:
If this step fails, is it safe to retry the delete? or will `h.Store.DeleteRetentionPolicy` fail on the second try because it is already deleted?

Reply:
This is the algorithm used in coordinator/statement_executor.go/executeDropRetentionPolicyStatement 

I wanted to maintain compatibility with whatever InfluxQL did.

Reply:
If we're trying to be the same, can we extract the method and reuse it?

Reply:
Not easily.  The two calls are the same and in the same order, but the error handling is `h.HttpError()` versus `return err`

Reply:
Let's put a comment in the code then, that this is intended to match the other method.